drivers/net/can/spi/mcp25xxfd/mcp25xxfd-core.c

```diff
@@ -2447,6 +2462,124 @@ mcp25xxfd_register_quirks(struct mcp25xxfd_priv *priv)
 		priv->devtype_data.quirks |= MCP25XXFD_QUIRK_HALF_DUPLEX;
 }
 
+static int mcp25xxfd_oscout_prepare(struct clk_hw *hw)
+{
+	struct mcp25xxfd_priv *priv =
+		container_of(hw, struct mcp25xxfd_priv, oscout);
+	struct net_device *ndev = priv->ndev;
+	u32 osc, osc_reference, osc_mask;
+	int err;
+	u8 mode;
+
+	/* increase the usage count, and disallow suspend */
+	pm_runtime_get_sync(ndev->dev.parent);
+
+	/* Read MCP25XXFD_REG_OSC as the first thing if we are
+	 * possibly in sleep mode. If we read anything else first,
+	 * we run the risk of not hitting the workaround in
+	 * mcp25xxfd-regmap.c
+	 */
+	err = regmap_read(priv->map_reg, MCP25XXFD_REG_OSC, &osc);
+	if (err)
+		return err;
+
+	if (osc & MCP25XXFD_REG_OSC_OSCDIS) {
+		osc = FIELD_PREP(MCP25XXFD_REG_OSC_CLKODIV_MASK, MCP25XXFD_REG_OSC_CLKODIV_10);
+		err = regmap_write(priv->map_reg, MCP25XXFD_REG_OSC, osc);
+		if (err)
+			return err;
+	}
+
+	osc_reference = MCP25XXFD_REG_OSC_OSCRDY;
+	osc_mask = MCP25XXFD_REG_OSC_OSCRDY | MCP25XXFD_REG_OSC_PLLRDY;
+
+	/* Wait for "Oscillator Ready" bit */
+	err = regmap_read_poll_timeout(priv->map_reg, MCP25XXFD_REG_OSC, osc,
+					(osc & osc_mask) == osc_reference,
+					10000, 1000000);
+	if (err)
+		return err;
+
+	err = regmap_update_bits(priv->map_reg, MCP25XXFD_REG_OSC,
+			MCP25XXFD_REG_OSC_CLKODIV_MASK, priv->clkodiv);
+	if (err)
+		return err;
+
+	err = mcp25xxfd_chip_set_mode(priv, MCP25XXFD_REG_CON_MODE_CONFIG);
+	if (err)
```

Comment:
I think this will explode if you set the clock _after_ ifup.

Reply:
You are right. It never bites me because the clocks are prepared at boot, but i'll change that too.

Reply:
What's even worse is the reset of the mcp25xxfd during ifup. I'm not sure what that means to the clock output. And if you use a divider != the reset default, the clock out pin would run with the wrong clock speed for some µs. It's best to avoid glitches on the clock line, that might cause random hard to debug problems if you do a ifidown ifup sequence to recover your CAN interface.... So from my point of view moving the "wake up from sleep and reset" and the "put into sleep mode" into PM is the way to go.